eth/api_backend.go

```diff
@@ -308,9 +309,15 @@ func (b *EthAPIBackend) GetPoolTransaction(hash common.Hash) *types.Transaction
 	return b.eth.txPool.Get(hash)
 }
 
+// GetTransaction retrieves the lookup along with the transaction itself associate
+// with the given transaction hash. A non-nil error will be returned if the
+// transaction is not found.
 func (b *EthAPIBackend) GetTransaction(ctx context.Context, txHash common.Hash) (*types.Transaction, common.Hash, uint64, uint64, error) {
-	tx, blockHash, blockNumber, index := rawdb.ReadTransaction(b.eth.ChainDb(), txHash)
-	return tx, blockHash, blockNumber, index, nil
+	lookup, tx, err := b.eth.blockchain.GetTransactionLookup(txHash)
+	if err != nil {
+		return nil, common.Hash{}, 0, 0, fmt.Errorf("tx is not existent or not indexed, %w", err)
```

Comment:
Note that this is a behavorial change for the `eth_getTransactionByHash` RPC, which previously returned `null` in this case.  Which has a couple ramifications that should be considered:

- I imagine fair number of upstream projects will be "surprised" by this RPC now potentially returning errors.  For example I'm pretty sure most wallets use this method to determine whether a newly broadcast tx has reached the public mempool.
- The error "leaks" configuration settings of the node (the index head and tail).

Reply:
Good point, @fjl do you have any suggestion about the RPC behavioral change?

Reply:
> For example I'm pretty sure most wallets use this method to determine whether a newly broadcast tx has reached the public mempool.

If the tx is in the txpool, then it will be found for sure,
If the tx is not in the txpool yet, the error will be returned instead of returning nil, as you said.

I will ask team about their opinions.

Reply:
I mean, it *is* an internal error from one perspective: the indexing is not done so the results are not reliable. From that perspective, the error message feels correct.

Tho I wouldn't add that much info since it's hard to interpret. Probably "transactions still indexing" or something along those lines is enough.

Reply:
One possible solution we could do is to make indexing part of the sync. I.e. after snap sync is "done", we could have a next phase with indexing, whilst eth_syncing does not return false.

Still, the quirk is that the current block currently kind of means that everything that's needed has been done. On the other hand, IMO a syncing node should not be just dropped into prod serving RPC.

Reply:
The main point is, we can't just return an error when the tx doesn't exist. We can return an error while we are still indexing, but the code here doesn't do that. If we decide to add an error during indexing, it would have to be identifiable by error code. This a decision affecting all API consumers and also proxy/LB implementations like Infura's. They all need to change their behavior to retry the lookup later or on a different node.

Reply:
I have fixed the code.

The error is only returned if the transactions are not fully indexed.

> it would have to be identifiable by error code

It's on my todo list.

Reply:
fixed